File: server/src/main/java/com/hyperlogix/server/domain/PLGNetwork.java
package com.hyperlogix.server.domain;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.List;
import java.util.stream.Collectors; // Import Collectors

@Data
@AllArgsConstructor
public class PLGNetwork implements Cloneable {
  List<Truck> trucks;
  List<Station> stations;
  List<Order> orders;
  List<Incident> incidents;
  List<Roadblock> roadblocks;

  public int getTrucksCapacity() {
    return trucks.stream().mapToInt(Truck::getMaxCapacity).sum();
  }

  public List<Order> getCalculatedOrders() {
    return orders.stream()
        .filter(order -> order.getStatus() == OrderStatus.CALCULATING)
        .collect(Collectors.toList());
  }

  @Override
  public PLGNetwork clone() {
    try {
      PLGNetwork cloned = (PLGNetwork) super.clone();
      cloned.trucks = this.trucks.stream().map(Truck::clone).collect(Collectors.toList());
      cloned.stations = this.stations.stream().map(Station::clone).collect(Collectors.toList());
      cloned.orders = this.orders.stream().map(Order::clone).collect(Collectors.toList());
      cloned.incidents = List.copyOf(this.incidents);
      cloned.roadblocks = List.copyOf(this.roadblocks);
      return cloned;
    } catch (CloneNotSupportedException e) {
      throw new AssertionError(); // Can't happen
    }
  }
}

File: client/src/domain/PLGNetwork.ts
import type { Order } from './Order'
import type { Station } from './Station'
import type { Truck } from './Truck'
import type { Point } from '../api'

interface Incident {
  id: string
}

interface Roadblock {
  start: string // ISO date-time string
  end: string // ISO date-time string
  blockedNodes: Point[]
}

export interface PLGNetwork {
  trucks: Truck[]
  stations: Station[]
  orders: Order[]
  incidents: Incident[]
  roadblocks: Roadblock[]
}
